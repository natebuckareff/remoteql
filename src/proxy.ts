const nothing = (): void => {};
const captureSymbol = Symbol('capture');

export type Capture =
  | { kind: 'var' }
  | { kind: 'get'; parent: Capture; property: string }
  | { kind: 'apply'; parent: Capture; args: Capture[] }
  | { kind: 'map'; parent: Capture; args: Capture[]; output: Capture }
  | ValueCapture;

export type ValueCapture =
  | { kind: 'value'; type: 'primitive'; value: unknown }
  | { kind: 'value'; type: 'array'; value: Capture[] }
  | { kind: 'value'; type: 'object'; value: Record<string, Capture> }
  | { kind: 'value'; type: 'function' };

export type Tracker = (() => void) & { readonly [captureSymbol]: Capture };

export function isTracker(value: unknown): value is Tracker {
  return typeof value === 'function' && value[captureSymbol] !== undefined;
}

export function getCapture(value: Tracker): Capture;
export function getCapture(value: unknown): Capture | undefined;
export function getCapture(value: unknown): Capture | undefined {
  if (isTracker(value)) {
    return value[captureSymbol];
  }
}

export function createTracker<T extends object = Tracker>(source: Capture): T {
  return new Proxy<T>(nothing as any, {
    get(_target, p, _receiver) {
      if (p === captureSymbol) {
        return source;
      }

      if (typeof p === 'symbol') {
        throw Error(`cannot serialize symbols`);
      }

      return createTracker({
        kind: 'get',
        parent: source,
        property: p,
      });
    },

    apply(_target, _thisArg, argArray) {
      if (source.kind === 'get') {
        if (source.property === 'map') {
          const callback = argArray[0];
          if (argArray.length === 1 && typeof callback === 'function') {
            const args: Tracker[] = [];
            for (let i = 0; i < Math.min(callback.length, 3); ++i) {
              args.push(createTracker({ kind: 'var' }));
            }
            const result = callback(...args);

            return createTracker({
              kind: 'map',
              parent: source,
              args: args.map(arg => getCapture(arg)),
              output: wrapValue(result),
            });
          }
        }
      }

      return createTracker({
        kind: 'apply',
        parent: source,
        args: argArray.map(arg => wrapValue(arg)),
      });
    },
  });
}

function wrapValue(value: unknown): Capture {
  if (isTracker(value)) {
    return getCapture(value);
  }

  if (typeof value === 'function') {
    return {
      kind: 'value',
      type: 'function',
    };
  } else if (typeof value === 'object') {
    if (value === null) {
      return {
        kind: 'value',
        type: 'primitive',
        value: null,
      };
    }

    if (Array.isArray(value)) {
      return {
        kind: 'value',
        type: 'array',
        value: value.map(x => wrapValue(x)),
      };
    } else {
      const wrapped: Record<PropertyKey, Capture> = {};
      for (const [prop, field] of Object.entries(value)) {
        wrapped[prop] = wrapValue(field);
      }
      return {
        kind: 'value',
        type: 'object',
        value: wrapped,
      };
    }
  } else {
    return {
      kind: 'value',
      type: 'primitive',
      value: value,
    };
  }
}
