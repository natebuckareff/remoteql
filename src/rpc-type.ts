/** @internal */
export const rpcImplSymbol: unique symbol = Symbol();

export class RpcImpl {
  [rpcImplSymbol]: true;
}

export type PlainPrimitive =
  | undefined
  | null
  | boolean
  | number
  | string
  | bigint;

export type IsPlainPrimitive<T> = T extends
  | null
  | boolean
  | number
  | string
  | bigint
  ? T
  : never;

export type ConstrainPrimitive<T> = IsPlainPrimitive<T>;

export type Resolved<T> = T extends Rpc<infer U> ? U : never;

/** @internal */
declare const rpcPromiseSymbol: unique symbol;

export type RpcPromiseMarker<T> = {
  readonly [rpcPromiseSymbol]: T;
};

export type Rpc<T> = RpcPromiseMarker<T> &
  ([T] extends [PlainPrimitive]
    ? ConstrainPrimitive<T>
    : T extends (...args: infer Args) => infer ReturnType
    ? RpcFunction<Args, ReturnType>
    : T extends readonly [...any]
    ? number extends T["length"]
      ? RpcArray<T[number]>
      : RpcTuple<T>
    : T extends RpcImpl
    ? RpcClient<T>
    : T extends object
    ? RpcObject<T>
    : void extends T
    ? void
    : never);

export type RpcFunction<Args extends any[], ReturnType> = (
  ...args: RpcParams<Args>
) => Rpc<Awaited<ReturnType>>;

export type RpcParams<Params extends any[]> = {
  [K in keyof Params]: Params[K] | Rpc<Params[K]>;
};

export type RpcClient<T extends RpcImpl> = {
  [K in keyof T]: T[K] extends (...args: infer Args) => infer ReturnType
    ? RpcFunction<Args, ReturnType>
    : never;
};

/** @internal */
declare const rpcArraySymbol: unique symbol;

export type RpcObject<T extends object> = T extends
  | Function
  | ReadonlyArray<any>
  ? never
  : { [K in keyof T]: Rpc<T[K]> };

export type RpcArray<E> = {
  [Index in Extract<keyof E[], number>]: Rpc<E | undefined>;
} & {
  at(index: number): Rpc<E>;
  map<const U>(
    callback: (value: Rpc<E>, index: Rpc<number>) => U
  ): U extends Rpc<infer E> ? Rpc<E[]> : Rpc<U[]>;
};

export type RpcTuple<T extends readonly unknown[]> = RpcArray<T[number]> & {
  [K in keyof T]: Rpc<T[K]>;
};
