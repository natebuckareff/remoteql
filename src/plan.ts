import { type Capture, getCapture, type ValueCapture } from './proxy.js';
import type { Rpc } from './rpc-type.js';
import { type Serialized, serialize } from './serialize.js';

export type Frame = Record<number, Operation | Block>;

export type Operation =
  | ['let', Serialized]
  | ['get', ...(number | string)[]]
  | ['call', (number | string)[], ...number[]]
  | ['map', (number | string)[], number];

export interface Block {
  params: number[];
  frame: Frame;
  outputs: number[];
}

class Value {
  constructor(private value: unknown) {}

  get(): unknown {
    return this.value;
  }
}

export class Ref {
  constructor(private index: number) {}

  get(): number {
    return this.index;
  }
}

export class Plan {
  private frame: Frame = {};
  private ids: Map<unknown, number> = new Map();
  private nextId = 0;
  private outputs: number[] = [];

  // biome-ignore lint/suspicious/noExplicitAny: should accept any type of rpc value
  constructor(values: Rpc<any>[]) {
    this._walk(values);
  }

  getFrame(): Frame {
    return this.frame;
  }

  getOutputs(): number[] {
    return this.outputs;
  }

  private _getNextId(): number {
    const id = this.nextId;
    this.nextId += 1;
    return id;
  }

  private _getIdOrNext(cap: Capture): number {
    let id = this.ids.get(cap);
    if (id === undefined) {
      id = this._getNextId();
      this.ids.set(cap, id);
    }
    return id;
  }

  private _getIdOf(cap: Capture): number {
    const id = this.ids.get(cap);
    if (id === undefined) {
      throw Error('capture id not found');
    }
    return id;
  }

  private _walk(values: Rpc<unknown>[]): void {
    for (const value of values) {
      const cap = getCapture(value);
      if (cap === undefined) {
        throw Error('invalid rpc promise');
      }
      this._walkCapture(cap, this.frame);
      const outputId = this._getIdOf(cap);
      this.outputs.push(outputId);
    }
  }

  private _walkCapture(cap: Capture, frame: Frame): number | Value {
    const rec = (cap: Capture, frame: Frame): number => {
      const value = this._walkCapture(cap, frame);

      if (value instanceof Value) {
        const id = this._getNextId();
        const ser = serialize(value.get());
        frame[id] = ['let', ser];
        return id;
      } else {
        return value;
      }
    };

    switch (cap.kind) {
      case 'var':
        return this._getIdOrNext(cap);

      case 'get': {
        const [targetCap, path] = compressGetPath(cap);
        const targetId = rec(targetCap, frame);
        const id = this._getIdOrNext(cap);
        const fullPath = [targetId, ...path];
        const op: Operation = ['get', ...fullPath];
        frame[id] = op;
        return id;
      }

      case 'apply': {
        const [target, path] = compressGetPath(cap.parent);
        const parentId = rec(target, frame);
        const argIds: number[] = [];
        for (const arg of cap.args) {
          const argId = rec(arg, frame);
          argIds.push(argId);
        }
        const id = this._getIdOrNext(cap);
        const fullPath = [parentId, ...path];
        const op: Operation = ['call', fullPath, ...argIds];
        frame[id] = op;
        return id;
      }

      case 'map': {
        const block: Block = {
          params: [],
          frame: {},
          outputs: [],
        };

        const [target, path] = compressGetPath(cap.parent);
        const parentId = rec(target, frame);
        const fullPath = [parentId, ...path];

        const blockId = this._getNextId();
        for (const arg of cap.args) {
          const argId = rec(arg, block.frame);
          block.params.push(argId);
        }

        const outputId = rec(cap.output, block.frame);
        block.outputs.push(outputId);
        frame[blockId] = block;

        if (fullPath[fullPath.length - 1] !== 'map') {
          throw Error('invalid map operation');
        }

        const truncatedPath = fullPath.slice(0, fullPath.length - 1);
        const mapId = this._getIdOrNext(cap);
        const mapOp: Operation = ['map', truncatedPath, blockId];
        frame[mapId] = mapOp;
        return mapId;
      }

      case 'value':
        return this._walkValue(cap, frame);
    }
  }

  private _walkValue(cap: ValueCapture, frame: Frame): Value {
    const unwrap = (value: number | Value): unknown => {
      if (value instanceof Value) {
        return value.get();
      } else {
        return new Ref(value);
      }
    };

    switch (cap.type) {
      case 'primitive':
        return new Value(cap.value);

      case 'array': {
        const output: unknown[] = [];
        for (const element of cap.value) {
          output.push(unwrap(this._walkCapture(element, frame)));
        }
        return new Value(output);
      }

      case 'object': {
        const output: Record<string, unknown> = {};
        for (const [prop, field] of Object.entries(cap.value)) {
          output[prop] = unwrap(this._walkCapture(field, frame));
        }
        return new Value(output);
      }

      case 'function':
        throw Error('invalid rpc value');
    }
  }
}

function compressGetPath(cap: Capture): [Capture, (number | string)[]] {
  const path: (number | string)[] = [];
  while (true) {
    if (cap.kind === 'get') {
      path.unshift(cap.property);
      cap = cap.parent;
    } else {
      return [cap, path];
    }
  }
}
