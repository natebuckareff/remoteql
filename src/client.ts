import { BatchScheduler } from './batch.js';
import { Plan } from './plan.js';
import type { Capture } from './proxy.js';

export class Client {
  public batch: BatchScheduler<Capture, unknown>;

  constructor() {
    this.batch = new BatchScheduler<Capture, unknown>(caps => {
      const plan = new Plan(caps);
      const frame = plan.getFrame();
      const outputs = plan.getOutputs();
      throw Error('todo');
    });
  }
}
