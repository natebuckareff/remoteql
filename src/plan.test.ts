import { expect, test } from 'vitest';
import { Client } from './client.js';
import { Plan } from './plan.js';
import { createTracker } from './proxy.js';
import { type Rpc, RpcImpl } from './rpc-type.js';

class Basic extends RpcImpl {
  async getUsers(): Promise<User[]> {
    throw Error('todo');
  }

  async getUserById(id: number): Promise<User> {
    throw Error('todo');
  }
}

interface User {
  id: number;
  name: string;
  friends: number[];
}

test('basic plan', () => {
  const client = new Client();
  const rpc = createTracker<Rpc<Basic>>(client, { kind: 'var' });
  const me = rpc.getUserById(42);
  const users = rpc.getUsers();
  const firstUser = users[0];
  const usersWithFriends = users.map(user => ({
    info: {
      id: user.id,
      friends: user.friends.map(id => rpc.getUserById(id)),
    },
  }));
  const op = new Plan([me, firstUser, usersWithFriends]);
  const frame = op.getFrame();
  expect(frame).toMatchSnapshot();
  expect(op.getOutputs()).toEqual([2, 4, 13]);
});
