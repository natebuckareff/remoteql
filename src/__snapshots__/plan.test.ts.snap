// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic plan 1`] = `
{
  "1": [
    "let",
    42,
  ],
  "13": [
    "map",
    [
      3,
    ],
    5,
  ],
  "2": [
    "call",
    [
      0,
      "getUserById",
    ],
    1,
  ],
  "3": [
    "call",
    [
      0,
      "getUsers",
    ],
  ],
  "4": [
    "get",
    3,
    "0",
  ],
  "5": {
    "frame": {
      "11": [
        "map",
        [
          6,
          "friends",
        ],
        8,
      ],
      "12": [
        "let",
        {
          "info": {
            "friends": [
              11,
            ],
            "id": [
              7,
            ],
          },
        },
      ],
      "7": [
        "get",
        6,
        "id",
      ],
      "8": {
        "frame": {
          "10": [
            "call",
            [
              0,
              "getUserById",
            ],
            9,
          ],
        },
        "outputs": [
          10,
        ],
        "params": [
          9,
        ],
      },
    },
    "outputs": [
      12,
    ],
    "params": [
      6,
    ],
  },
}
`;
